### ISSUE TITLE
ReadNode throw exception

### ISSUE THREAD

**Issue by verysimplenick** [KirillOsenkov/MSBuildStructuredLog#82]
Hi, I try build my solution with your tool, but catch exception. Builded with vs 2017.2 normally. (MSBuild15)
![ExceptionInfo](https://user-images.githubusercontent.com/922660/27908446-8c52d272-6254-11e7-9d7a-6f51499ddb5f.png)

```txt
System.ArgumentNullException: Значение не может быть неопределенным.
Имя параметра: key
   в System.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   в System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
   в System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
   в Microsoft.Build.Logging.StructuredLogger.Serialization.CreateNode(String name)
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.ReadNode()
   в Microsoft.Build.Logging.StructuredLogger.BinaryLogReader.Read(String filePath)
   в Microsoft.Build.Logging.StructuredLogger.Serialization.Read(String filePath)
   в StructuredLogViewer.HostedBuild.<>c__DisplayClass6_0.<BuildAndGetResult>b__0() в C:\MSBuildStructuredLog\src\StructuredLogViewer\HostedBuild.cs:строка 64
```

**Comment by verysimplenick** [comment]
After open your tool with admin rights everything works ok.
Maybe you need catch `SecurityException` and `UnauthorizedAccessException` to show more readable description of error.

**Comment by KirillOsenkov** [comment]
Hmm, this is very weird, from the call stack it more looks like a corrupted file than permissions. It's already reading the file so it must have permissions to read it. Can you reproduce it? If so, while it's building, could you open the %TEMP% directory, sort the files by date (most recent first) and save the most recent .buildlog file? Then try opening that file in the viewer and see if you get the same error. Ideally if you get the error if you could send me that file? kirill dot osenkov at microsoft dot com.

Also I wonder if it's happening just for that particular solution or for other solutions as well?

**Comment by verysimplenick** [comment]
I tried reproduce, but it's was happened not often. Finally I reproduce this, send `.buildlog` to your email. Try open it.

**Comment by KirillOsenkov** [comment]
Thanks, I'm able to reproduce with the file you've sent me, it is indeed pretty corrupted.

I have a suspicion that the problem is that you're using localized (Russian) MSBuild and this tool currently only supports English MSBuild. I knew it would be broken, but I didn't think it would be so broken.

The thing that surprises me the most is that it sometimes works for you. I don't understand how this is possible yet.

Sorry about the inconvenience. One day I'll get to supporting localized MSBuild but unfortunately I won't have time for this any time soon.

**Comment by KirillOsenkov** [comment]
Note to self: saved the corrupted log file privately under Work\Bugs\CorruptedMSBuildLog-Issue82.

**Comment by verysimplenick** [comment]
Yep, sometimes everything is ok! :) Kirill, what about localized 2017.3 msbuild + binlog, this problem can be reproduced on them?

**Comment by KirillOsenkov** [comment]
I'm guessing the problem will reproduce when you're trying to open the .binlog file in the viewer, because internally it has to first convert to .buildlog and this is what's broken. The .binlog itself should be OK and openable on an English locale. If you have 15.3 already, could you send me a .binlog file so I could verify? Also from the case when .buildlog opens successfully, could you send me the log that opens fine? Just so I can compare? Thanks for your help!

**Comment by verysimplenick** [comment]
Can reproduce at VS 2017.3 + latest MSBuildStructuredLog
[error.zip](https://github.com/KirillOsenkov/MSBuildStructuredLog/files/1348705/error.zip)

**Comment by KirillOsenkov** [comment]
Can you try getting a .binlog file using http://msbuildlog.com (basically run `msbuild.exe /bl from command-line`)

And just send me the .binlog file.

**Comment by japj** [comment]
Hi, we also seem to run into this with v2.0.81
It seems to reproduce when there are certain kind of build failures (that we are trying to analyze).
Not sure yet what the exact scenario is to reproduce this in a clean/isolated way for you.

This is with running the actual project from MSBuild Structured Log Viewer (through "Build Solution")
```
System.ArgumentNullException: Value cannot be null.
Parameter name: key
   at System.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
   at System.Collections.Generic.Dictionary`2.FindEntry(TKey key)
   at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
   at Microsoft.Build.Logging.StructuredLogger.Serialization.CreateNode(String name)
   at Microsoft.Build.Logging.StructuredLogger.BuildLogReader.ReadNode()
   at Microsoft.Build.Logging.StructuredLogger.BuildLogReader.ReadNode()
   at Microsoft.Build.Logging.StructuredLogger.BuildLogReader.ReadNode()
   at Microsoft.Build.Logging.StructuredLogger.BuildLogReader.ReadNode()
   at Microsoft.Build.Logging.StructuredLogger.BuildLogReader.ReadNode()
   at Microsoft.Build.Logging.StructuredLogger.BuildLogReader.ReadNode()
   at Microsoft.Build.Logging.StructuredLogger.BuildLogReader.ReadNode()
   at Microsoft.Build.Logging.StructuredLogger.BuildLogReader.ReadNode()
   at Microsoft.Build.Logging.StructuredLogger.BuildLogReader.ReadNode()
   at Microsoft.Build.Logging.StructuredLogger.BuildLogReader.Read(Stream stream, Byte[] projectImportsArchive)
   at Microsoft.Build.Logging.StructuredLogger.BuildLogReader.Read(String filePath)
   at Microsoft.Build.Logging.StructuredLogger.Serialization.Read(String filePath)
   at StructuredLogViewer.HostedBuild.<>c__DisplayClass5_0.<BuildAndGetResult>b__0()
```

**Comment by KirillOsenkov** [comment]
@japj any chance you could share the corrupt .binlog with me privately? If that's not an option, would you consider cloning this repo, building in Debug yourself and investigating what is going on? Thanks!

**Comment by KirillOsenkov** [comment]
Actually that might be a bug in the .buildlog format (internal format of the Viewer, used to serialize the tree). Can you try building that project from command line using msbuild `/bl` and then opening the resulting .binlog? I wonder if that would work fine? If you can, send me the .binlog and the .buildlog files.

**Comment by japj** [comment]
I tried to reproduce it earlier today but msbuild from cmdline actually succeeded and I did not have time yet to isolate a scenario for reproducing. In the mean time other things got fixed in the code so i need to go back and figure out exactly which changeset still contains the issue to generate binlog files for analysis.

Not entirely sure, but I think the problem has at least occurred during an error of a XamlTaskFactory command.

I tried saving the ‘corrupt’ binlog from the menu earlier, but that ended with an empty binlog and I just realized the viewer probably stores it in a temp folder when running solution build?

Is there a (Microsoft) email address I can privately send binlog files when I have them?
